sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedPlainValuesReader.java

```diff
@@ -53,19 +53,47 @@ public void skip() {
     throw new UnsupportedOperationException();
   }
 
+  private void updateCurrentByte() {
+    try {
+      currentByte = (byte) in.read();
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Failed to read a byte", e);
+    }
+  }
+
   @Override
   public final void readBooleans(int total, WritableColumnVector c, int rowId) {
-    // TODO: properly vectorize this
-    for (int i = 0; i < total; i++) {
-      c.putBoolean(rowId + i, readBoolean());
+    int i = 0;
+    if (bitOffset > 0) {
+      i = Math.min(8 - bitOffset, total);
```

Comment:
Do we need to update the total here?

Reply:
I don't think we need to update the `total` here. Only the `I` gets updated.
Any particular examples that you are worried?

Reply:
Never mind, I missed that you were updating the variable i.

By the way, what happens when total = 8 - bitOffset? Do we need to call updateCurrentByte() somewhere for this as bitOffset will become 0?

Reply:
We do not need to call `updateCurrentByte()` here when `total == 8 - bitOffset`. If you scroll down to the one bit version of the reader in the same file, you will find 
```
  public final boolean readBoolean() {
    if (bitOffset == 0) {
      updateCurrentByte();
    }
```
It calls `updateCurrentByte` right before reading. So it is always next reader's responsibility to call the method.

Reply:
Yes, I understand but what about `readBooleans()` method? Does it mean that I can only call readBooleans() once?

Reply:
`readBooleans()` can be called multiple times. The scenario is tested in `ColumnarBatchSuite.scala`.

`readBooleans()` also calls `updateCurrentByte()` right before using it.
```
public final void readBooleans(...
    if (bitOffset > 0) { // means there are still bits to be read in currentByte
...
    for (; i + 7 < total; i += 8) {
      updateCurrentByte(); // calling it here
...
    if (i < total) {
      updateCurrentByte(); // calling it here
```

Reply:
Does the code work for total = 9 and bitOffset = 0?

Reply:
The code work for `total = 9` and `bitOffset = 0`. 

```
public final void readBooleans(...
    if (bitOffset > 0) { // it will not enter here
...
    for (; i + 7 < total; i += 8) {
      updateCurrentByte(); // the whole one byte (8bits) is read here
...
    if (i < total) {
      updateCurrentByte(); // the last 1 bit is read here
```

There is already a similar test for `total=8`, `bitOffset=1`. But I added `total = 9` and `bitOffset = 0` too.